Update Matter Casting Mobile App README documentation (#177)
sharadb-amazon/connectedhomeip: examples/tv-casting-app/APIs.md

# Matter Casting APIs

Matter Casting consists of three parts:

> * **The mobile app**: For most content providers, this would be your consumer-facing mobile app.  By making your moile app a Matter "Casting Client", you enable the user to discover casting targets, cast content, and control casting sessions.

Reviewer: ```suggestion
* **The mobile app**: For most content providers, this would be your consumer-facing mobile app.  By making your moile app a Matter "Casting Client", you enable the user to discover casting targets, cast content, and control casting sessions. The [example Matter tv-casting-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-casting-app) for Android / iOS and Linux builds on top of the Matter SDK to demonstrate how a TV Casting mobile app works.
```

> * **The mobile app**: For most conent providers, this would be your consumer-facing mobile app.  By adding a Matter "Casting Client" to your mobile app, you enable the user to discover casting targets, cast content, and control casting sessions.

Reviewer: ```suggestion
* **The mobile app**: For most content providers, this would be your consumer-facing mobile app.  By making your moile app a Matter "Casting Client", you enable the user to discover casting targets, cast content, and control casting sessions.
```

> * **The TV content app**: For most conent providers, this would be your consumer-facing app on a Smart TV.  By enhancing your TV app to act as a Matter "Casting Player", you enable Matter Casting Clients like your mobile app to cast content. The [example Matter tv-casting-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-casting-app) for Android / iOS and Linux builds on top of the Matter SDK to demonstrate how a TV app works.

Reviewer: ```suggestion
* **The TV content app**: For most content providers, this would be your consumer-facing app on a Smart TV.  By enhancing your TV app to act as a Matter "Content app", you enable Matter Casting Clients to cast content. The [example Matter content-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-app/android/App/content-app) for Android builds on top of the Matter SDK to demonstrate how a TV Content app works.
```

> * **The TV platform app**: The TV platform app is generally implemented by the TV manufacturer and provides common capabilities around media playback on the TV like the Basic Video Player.

Reviewer: ```suggestion
* **The TV platform app**: The TV platform app is generally implemented by the TV manufacturer and provides common capabilities around media playback on the TV like the Basic Video Player. The [example Matter tv-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-app) for Android builds on top of the Matter SDK to demonstrate how a TV platform app works.
```

The steps to start a casting session are:

1. [Initialize](#initialize-the-casting-client) the `CastingClient` using the Matter SDK.
1. [Discover](#discover-casting-players) `CastingPlayer` devices using Matter Commissioner discovery.
1. [Connect](#connect-to-a-casting-player) to the `CastingPlayer` to discover available endpoints.
1. [Select](#select-an-endpoint-on-the-casting-player) an available `Endpoint` hosted by the `CastingPlayer`.

Next, you're ready to:

1. [Issue commands](#issuing-commands) to the `Endpoint`.

> 1. [Read](#read-operations) playback state.

Reviewer: This "playback state" feels like a fuzzy term, given that there is a specific attribute in the MediaPlayback cluster called CurrentState. But a Casting client may be interested in reading many different attributes beyond this one.

Author: I'm open to suggestions here... maybe we can say "Read endpoint attributes like playback state"?

Reviewer: Yes, that'll make more sense!

1. [Subscribe](#subscriptions) to playback events.

## Build and Setup
The Casting Client is expected to consume the Matter TV Casting library built for its respective platform to consume any of the APIs described in this document. Refer to the tv-casting-app READMEs for [Android](android) and [iOS](darwin) to understand how to build and consume each platform's specicific libraries.

### Initialize the Casting Client
_{Complete Initialization examples: [Android](android/App/app/src/main/java/com/matter/casting/InitializationExample.java) |  [iOS](darwin/TvCasting/TvCasting/TvCastingApp.swift)}_

> A Casting Client must first define the following [`DataProvider`](android/App/app/src/main/jni/com/matter/casting/params/DataProvider.java) objects for the the Matter Casting library to use throughout the client's lifecycle:

Reviewer: Can we also use the word "initialize" to indicate the this also initializes the whole SDK i.e. they are not just setting these DataProvider params.

Author: I just committed a change to tweak the wording.

    }  
};  
```

> On iOS, implement a `func addDiscoveredCastingPlayers`, `func removeDiscoveredCastingPlayers` and `func updateDiscoveredCastingPlayers` which to listen to notifications as Casting Players are added, removed, or updated.

Reviewer: ```suggestion
On iOS, implement a `func addDiscoveredCastingPlayers`, `func removeDiscoveredCastingPlayers` and `func updateDiscoveredCastingPlayers` which listen to notifications as Casting Players are added, removed, or updated.
```

On iOS, register the listeners by calling `addObserver` on the `NotificationCenter` with the appropriate selector, and then call `start` on the `sharedInstance` of `MTRCastingPlayerDiscovery`.
```objectivec
func startDiscovery() {
    NotificationCenter.default.addObserver(
    self,
    selector: #selector(self.addDiscoveredCastingPlayers),
    name: NSNotification.Name.didAddCastingPlayers,
    object: nil)
    NotificationCenter.default.addObserver(
    self,
    selector: #selector(self.removeDiscoveredCastingPlayers),
    name: NSNotification.Name.didRemoveCastingPlayers,
    object: nil)
    NotificationCenter.default.addObserver(
    self,
    selector: #selector(self.updateDiscoveredCastingPlayers),
    name: NSNotification.Name.didUpdateCastingPlayers,

>     object: nil)

Reviewer: nit: I thought the single liner NotificationCenter statements were easier to read. You can keep the new formatting if you did it deliberately, just thought I'll call out that it looks odd to me.

Author: I don't have a strong opinion here and am fine if you want to revert it.  It just went to a long line that was inconsistent with the line lengths of other samples, requiring left-->right scrolling.
